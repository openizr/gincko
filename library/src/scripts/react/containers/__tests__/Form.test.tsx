/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { act, Simulate } from 'react-dom/test-utils';
import Form from 'scripts/react/containers/Form';
import { render, unmountComponentAtNode } from 'react-dom';

jest.mock('scripts/core/Engine');
jest.mock('scripts/react/components/Step', () => ({ onUserAction }: Json): JSX.Element => {
  onUserAction();
  return <div id="Step" />;
});
jest.mock('sonar-ui/react', () => ({
  generateRandomId: (): string => '_abcde',
  markdown: (value: string): string => value,
  buildClass: (...values: string[]): string => values.join(' '),
}));
jest.mock('diox/connectors/react', () => jest.fn(() => [jest.fn(() => [{
  steps: [{ id: 'start' }, { id: 'end' }],
  loadingNextStep: process.env.LOADING === 'true',
}]), jest.fn()]));

describe('containers/Form', () => {
  let container = document.createElement('div');
  const customComponents = {};

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    jest.clearAllMocks();
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    ((container as unknown) as null) = null;
  });

  test('loading next step', () => {
    process.env.LOADING = 'true';
    const preventDefault = jest.fn();
    act(() => {
      render(<Form
        configuration={{
          root: 'start',
          steps: {
            start: {
              fields: [],
            },
          },
          fields: {},
        }}
        customComponents={customComponents}
      />, container);
    });
    act(() => {
      Simulate.submit(document.querySelector('form') as HTMLFormElement, { preventDefault });
    });
    delete process.env.LOADING;
    expect(preventDefault).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test('with active step', () => {
    act(() => {
      render(<Form
        configuration={{
          root: 'start',
          steps: {
            start: {
              fields: [],
            },
          },
          fields: {},
        }}
        activeStep="start"
        customComponents={customComponents}
      />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('not loading next step', () => {
    act(() => {
      render(<Form
        configuration={{
          root: 'start',
          steps: {
            start: {
              fields: [],
            },
          },
          fields: {},
        }}
        activeStep="start"
        customComponents={customComponents}
      />, container);
    });
    expect(container).toMatchSnapshot();
  });
});
