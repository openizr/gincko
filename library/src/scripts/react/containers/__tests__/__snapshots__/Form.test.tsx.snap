// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Form loading next step 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="form__steps"
    >
      <div
        id="Step"
      />
      <div
        id="Step"
      />
      <div
        class="ui-loader"
      />
    </div>
  </form>
</div>
`;

exports[`containers/Form not loading next step 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="form__steps"
    >
      <div
        id="Step"
      />
      <div
        id="Step"
      />
    </div>
  </form>
</div>
`;

exports[`containers/Form with active step 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="form__steps"
    >
      <div
        id="Step"
      />
      <div
        id="Step"
      />
    </div>
  </form>
</div>
`;
